crates/rolldown_plugin_dynamic_import_vars/src/lib.rs

```diff
@@ -0,0 +1,169 @@
+use clone_expr::clone_expr;
+use oxc::{
+  ast::{
+    ast::{Expression, ImportOrExportKind, Statement, TSTypeParameterInstantiation},
+    AstBuilder, VisitMut,
+  },
+  span::{Span, SPAN},
+  syntax::number::NumberBase,
+};
+use rolldown_plugin::{
+  HookLoadArgs, HookLoadOutput, HookLoadReturn, HookResolveIdArgs, HookResolveIdOutput,
+  HookResolveIdReturn, HookTransformAstArgs, HookTransformAstReturn, Plugin, SharedPluginContext,
+};
+use std::borrow::Cow;
+use to_glob::to_glob_pattern;
+mod clone_expr;
+mod should_ignore;
+mod to_glob;
+
+const DYNAMIC_IMPORT_HELPER: &str = "\0rolldown_dynamic_import_helper.js";
+
+#[derive(Debug)]
+pub struct DynamicImportVarsPlugin {}
+
+impl Plugin for DynamicImportVarsPlugin {
+  fn name(&self) -> Cow<'static, str> {
+    Cow::Borrowed("dynamic_import_vars")
+  }
+
+  async fn resolve_id(
+    &self,
+    _ctx: &SharedPluginContext,
+    args: &HookResolveIdArgs<'_>,
+  ) -> HookResolveIdReturn {
+    if args.specifier == DYNAMIC_IMPORT_HELPER {
+      Ok(Some(HookResolveIdOutput { id: DYNAMIC_IMPORT_HELPER.to_string(), ..Default::default() }))
+    } else {
+      Ok(None)
+    }
+  }
+
+  async fn load(&self, _ctx: &SharedPluginContext, args: &HookLoadArgs<'_>) -> HookLoadReturn {
+    if args.id == DYNAMIC_IMPORT_HELPER {
+      Ok(Some(HookLoadOutput {
+        code: include_str!("dynamic_import_helper.js").to_string(),
+        ..Default::default()
+      }))
+    } else {
+      Ok(None)
+    }
+  }
+
+  fn transform_ast(
+    &self,
+    _ctx: &SharedPluginContext,
+    mut args: HookTransformAstArgs,
+  ) -> HookTransformAstReturn {
+    // TODO: Ignore if includes a marker like "/* @rolldown-ignore */"
+    args.ast.program.with_mut(|fields| {
+      let ast_builder: AstBuilder = AstBuilder::new(fields.allocator);
+      let mut visitor = DynamicImportVarsVisit { ast_builder, need_helper: false };
+      visitor.visit_program(fields.program);
+      if visitor.need_helper {
+        fields.program.body.push(visitor.import_helper());
+      }
+    });
+    Ok(args.ast)
+  }
+}
+
+pub struct DynamicImportVarsVisit<'ast> {
+  ast_builder: AstBuilder<'ast>,
+  need_helper: bool,
+}
+
+impl<'ast> VisitMut<'ast> for DynamicImportVarsVisit<'ast> {
+  #[allow(clippy::too_many_lines)]
+  fn visit_expression(&mut self, expr: &mut Expression<'ast>) {
+    if let Expression::ImportExpression(import_expr) = expr {
+      let pattern = to_glob_pattern(&import_expr.source).unwrap();
+      if let Some(pattern) = pattern {
+        self.need_helper = true;
+        *expr = self.call_helper(import_expr.span, pattern.as_str(), &import_expr.source);
+      }
+    }
+  }
+}
+
+impl<'ast> DynamicImportVarsVisit<'ast> {
+  /// generates:
+  /// ```js
+  /// __variableDynamicImportRuntimeHelper((import.meta.glob(pattern)), expr, segs)
+  /// ```
+  fn call_helper(&self, span: Span, pattern: &str, expr: &Expression<'ast>) -> Expression<'ast> {
```

Comment:
Look like here need to support query, it need to add some unit test for https://github.com/vitejs/vite/blob/main/packages/vite/src/node/__tests__/plugins/dynamicImportVar/parse.spec.ts.

Reply:
Some of the tests are now added in a modified form. This is because the raw import expression (in AST form) cannot be "resolved" as a path like what is done in Vite. Here the import expression is firstly converted to glob pattern, then be split into path and query.